xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -0,0 +1,425 @@
+    FaultConfig.FractionalPercent fractionalPercent = faultAbort.percent();
+    if (faultAbort.headerAbort()) {
+      try {
+        String grpcCodeStr = headers.get(HEADER_ABORT_GRPC_STATUS_KEY);
+        if (grpcCodeStr != null) {
+          int grpcCode = Integer.parseInt(grpcCodeStr);
+          abortStatus = Status.fromCodeValue(grpcCode);
+        }
+        String httpCodeStr = headers.get(HEADER_ABORT_HTTP_STATUS_KEY);
+        if (httpCodeStr != null) {
+          int httpCode = Integer.parseInt(httpCodeStr);
+          abortStatus = GrpcUtil.httpStatusToGrpcStatus(httpCode);
+        }
+        String abortPercentageStr = headers.get(HEADER_ABORT_PERCENTAGE_KEY);
+        if (abortPercentageStr != null) {
+          int abortPercentage =
+              Integer.parseInt(headers.get(HEADER_ABORT_PERCENTAGE_KEY));
+          if (abortPercentage >= 0 && abortPercentage < fractionalPercent.numerator()) {
+            fractionalPercent = FaultConfig.FractionalPercent.create(
+                abortPercentage, fractionalPercent.denominatorType());
+          }
+        }
+      } catch (NumberFormatException e) {
+        return null; // treated as header_abort not applicable
+      }
+    } else {
+      abortStatus = faultAbort.status();
+    }
+    if (random.nextInt(1_000_000) >= getRatePerMillion(fractionalPercent)) {
+      return null;
+    }
+    return abortStatus;
+  }
+
+  private static int getRatePerMillion(FaultConfig.FractionalPercent percent) {
+    int numerator = percent.numerator();
+    FaultConfig.FractionalPercent.DenominatorType type = percent.denominatorType();
+    switch (type) {
+      case TEN_THOUSAND:
+        numerator *= 100;
+        break;
+      case HUNDRED:
+        numerator *= 10_000;
+        break;
+      case MILLION:
+      default:
+        break;
+    }
+    if (numerator > 1_000_000 || numerator < 0) {
+      numerator = 1_000_000;
+    }
+    return numerator;
+  }
+
+  /** A {@link DelayedClientCall} with a fixed delay. */
+  private final class DelayInjectedCall<ReqT, RespT> extends DelayedClientCall<ReqT, RespT> {
+    final Object lock = new Object();
+    ScheduledFuture<?> delayTask;
+    boolean cancelled;
+
+    DelayInjectedCall(
+        long delayNanos, Executor callExecutor, ScheduledExecutorService scheduler,
+        @Nullable Deadline deadline,
+        final Supplier<? extends ClientCall<ReqT, RespT>> callSupplier) {
+      super(callExecutor, scheduler, deadline);
+      activeFaultCounter.incrementAndGet();
+      ScheduledFuture<?> task = scheduler.schedule(
+          new Runnable() {
+            @Override
+            public void run() {
+              synchronized (lock) {
+                if (!cancelled) {
+                  activeFaultCounter.decrementAndGet();
+                }
+              }
+              setCall(callSupplier.get());
+            }
+          },
+          delayNanos,
+          NANOSECONDS);
+      synchronized (lock) {
```

Comment:
What does the following lines of code do? How can this be cancelled when it is instantiating? Why cancel itself?

Reply:
The instantiation in `super()` may trigger an immediate concurrent cancellation if deadline is exceeded. If that happens, there could be a race, need to guarantee the task is cancelled.

Reply:
Why the task is unconditionally cancelled here?

Reply:
It's conditional, only cancel the task if  `cancelled` is true.

Reply:
Ohhhhh. Sorry, there is a `return` there... This is opposite from the existing implementation...